rfcs/reorganize_src_lib.md

```diff
@@ -0,0 +1,117 @@
+
+- The `crypto` folder (at least the libraries which were part of [this PR](https://github.com/MinaProtocol/mina/pull/9540)), and I see that this was already started in the
+`feature/group-lookups-mina` branch.
+
+## First attempt at grouping
+
+
+I tried to create meaningful groups and to not have to many edges in the following graph of [dependencies between groups](./res/reorganize_src_lib/class_graph.png) (which is still a bit complex).
+
+- The section below contains some information about the groups.
+- In the following dependency graphs, the libraries are colored according to their respective groups. The first one is the most readable (but contains less information) because it does not contain the `Utils` and `Logging` groups, and was run through the `tred` tool, which removes a lot of edges while preserving reachability.
+
+    - [without utils and logging - treded](./res/reorganize_src_lib/no_utils_tred.png)
+    - [without utils and logging](./res/reorganize_src_lib/no_utils.png)
+    - [all libraries - treded](./res/reorganize_src_lib/all_libs_tred.png)
+    - [all libraries](./res/reorganize_src_lib/all_libs.png)
+
+- The exact groups can also be found in the [library_group.md](./res/reorganize_src_lib/library_groups.md) file.
+
+
+
+### Current Groups
+
+- `Others`: Libraries that could stay at the toplevel, as their own group, or that I am not sure about.
+For instance the `mina_base` library that contains a lot a files and will maybe be split in the future.
+
+- `Utils`: Non application specific and usually small libraries. Most of the time leaves of the dependency graph.
+
+- `Logging`: Libraries related to logging that everyone can depend on (even the `Utils` one).
+
+- `Testing`: This folder would would contain two types of testing libraries:
+    - High level calls to tests such as `command_line_tests`.
+    - Test related utils such as `quickcheck_lib` (The difference with the `Utils` folder is that these would be allowed to depend on other application specific libraries).
+
+  The other folders could also contain a `testing` subfolder for specific test related libraries (for instance the `mina_net2_tests` would be in the `Network` group).
+    
+
+- `Processes`: Libraries related to parallelism and inter process communication,
+   such as `child_processes` and `pipe_lib`.
+
+- `Network`: Libraries related to network communication between the various component.
+
+- `Config`: Libraries such as `mina_numbers` defining primitive types used by most other groups.
+
+- `Crypto`:
+  Creating the `crypto` folder was already attempted in [this PR](https://github.com/MinaProtocol/mina/pull/9540) that I used as a reference.
+  Would it make sense to split it between the libraries that depend on the `Config` group,
+  and those that do not ?
+
+- `Snarky`: The libraries from the snarky git submodule.
+
+- `Transition`: Modules related to the handling of transitions
```

Comment:
I gave a try at separating the type and logic of the `mina_transition` lib, and noticed that it is quite easy to introduce unneeded dependencies after something like this, especially since the `implicit_transitive_deps` option of dune is `true`.
(Unneeded dependencies may provoke unnecessary rebuilds, missed opportunities for parallelism, and they might interfere with the reorganization of the `src/lib` folder).

### `implicit_transitive_deps`
Setting `implicit_transitive_deps` to `false` has two advantages:

- it does not prevent unneeded dependencies but makes it easier to check for them, because we can try to remove a dependency and see if the build still works (without it, the library may be needed but still reachable).
- we have more precise information in the  `dune describe` output for dependency graphs.

Was this already attempted? I see that there is an issue about it: https://github.com/MinaProtocol/mina/issues/2451, so maybe they are some known problems ?

### Unneeded dependencies

Even if `implicit_transitive_deps` is set, there could be some superfluous dependencies.

I think I found some of them by querying the dependency graphs and using info from `codept`.
See [this commit](https://github.com/MinaProtocol/mina/commit/8126856fea8e2735dbfa802d65e729334da8dbee) that removes these dependencies (I am using the nix setup for my work environment so I am a working on the `tweag-cleanup` branch for now).


This commit only removes edges to libraries that are not reachable by another path.
I focused on these because they are more interesting in term of build performance, and because it is easier to test that they are indeed unneeded: if the build work without these edges, it is not because the dependencies are reachable by another path.

The method I used to find these unneeded dependencies was quick but is quite hacky at the moment (in particular it ignores generated files and ppx). It could be cleaned up and made more correct but I wonder if it is worthwhile because there is an open dune issue about this (https://github.com/ocaml/dune/issues/650), so it should not be a problem in the future.